kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/Protocol.java

```diff
@@ -96,32 +118,392 @@ public String toString() {
+   *
+   * <p>Examples regarding feature status:
+   *
+   * <ul>
+   *   <li>from `[appendOnly]` to `[appendOnly]` => allowed.
+   *   <li>from `[appendOnly, changeDataFeed]` to `[appendOnly]` => not allowed.
+   * </ul>
+   */
+  public boolean canUpgradeTo(Protocol to) {
+    return to.getImplicitlyAndExplicitlySupportedFeatures()
+        .containsAll(this.getImplicitlyAndExplicitlySupportedFeatures());
+  }
+
+  /**
+   * Protocol normalization is the process of converting a table features protocol to the weakest
+   * possible form. This primarily refers to converting a table features protocol to a legacy
+   * protocol. A Table Features protocol can be represented with the legacy representation only when
+   * the features set of the former exactly matches a legacy protocol.
+   *
+   * <p>Normalization can also decrease the reader version of a table features protocol when it is
+   * higher than necessary.
+   *
+   * <p>For example:
+   *
+   * <ul>
+   *   <li>(1, 7, AppendOnly, Invariants, CheckConstraints) -> (1, 3)
+   *   <li>(3, 7, RowTracking) -> (1, 7, RowTracking)
+   * </ul>
+   */
+  public Protocol normalized() {
+    // Normalization can only be applied to table feature protocols.
+    if (!supportsWriterFeatures) {
+      return this;
+    }
+
+    Tuple2<Integer, Integer> versions =
+        TableFeatures.minimumRequiredVersions(getExplicitlySupportedFeatures());
+    int minReaderVersion = versions._1;
+    int minWriterVersion = versions._2;
+    Protocol newProtocol = new Protocol(minReaderVersion, minWriterVersion);
+
+    if (this.getImplicitlyAndExplicitlySupportedFeatures()
+        .equals(newProtocol.getImplicitlyAndExplicitlySupportedFeatures())) {
+      return newProtocol;
+    } else {
+      // means we have some feature that is added after table feature support.
+      // Whatever the feature (reader or readerWriter), it is always going to
+      // have minWriterVersion as 7. Required minReaderVersion
+      // should be based on the supported features.
+      return new Protocol(minReaderVersion, TABLE_FEATURES_MIN_WRITER_VERSION)
+          .withFeatures(getExplicitlySupportedFeatures());
+    }
+  }
+
+  /**
+   * Protocol denormalization is the process of converting a legacy protocol to the equivalent table
+   * features protocol. This is the inverse of protocol normalization. It can be used to allow
+   * operations on legacy protocols that yield results which cannot be represented anymore by a
+   * legacy protocol. For example
+   *
+   * <ul>
+   *   <li>(1, 3) -> (1, 7, readerFeatures=[], writerFeatures=[appendOnly, invariants,
+   *       checkConstraints])
+   *   <li>(2, 5) -> (2, 7, readerFeatures=[], writerFeatures=[appendOnly, invariants,
+   *       checkConstraints, changeDataFeed, generatedColumns, columnMapping])
+   * </ul>
+   */
+  public Protocol denormalized() {
+    // Denormalization can only be applied to legacy protocols.
+    if (supportsWriterFeatures) {
```

Comment:
I think it would be nice if we had a separate method (with some documentation) that explained the relationship between supportsWriterFeatures and is-legacy-protocol?

`private def isLegacyProtocol() = supportsWriterFeatures`

This would make it clearer in the code, and then you could also add some detail/comments in one spot that explaiend why?

Reply:
added utilities